daffodil-io/src/main/scala/org/apache/daffodil/io/DirectOrBufferedDataOutputStream.scala

```diff
@@ -33,81 +34,82 @@ import org.apache.daffodil.util.Maybe
 import org.apache.daffodil.util.Maybe._
 import org.apache.daffodil.util.MaybeULong
 import org.apache.daffodil.util.Misc
+import org.apache.daffodil.util.Logging
 import java.io.File
 
 /**
  * This simple extension just gives us a public method for access to the underlying byte array.
  * That way we don't have to make a copy just to access the bytes.
  */
-private[io] class ByteArrayOrFileOutputStreamWithGetBuf(
-  chunkSizeInBytes: Long = 4096,
-  maxBufferSizeInBytes: Long = 2000 * (1 << 20))
-  extends java.io.ByteArrayOutputStream {
-  private var _byteArrayOS = this
-  private var _fileOS: java.io.FileOutputStream = null
-
-  var isFile: Boolean = false
+private[io] class ByteArrayOutputStreamWithGetBuf() extends java.io.ByteArrayOutputStream {
+  def getCount = count
+  def getBuf = buf
+}
+
+private[io] class ByteArrayOrFileOutputStream(
+  maxBufferSizeInBytes: Long,
+  tempFilePath: Path,
+  useFileOnly: Boolean)
+  extends java.io.OutputStream
+  with Logging {
+  val filePath = {
+    if (tempFilePath.toString == "")
+      Paths.get(System.getProperty("java.io.tmpdir"))
+    else
+      tempFilePath
+  }
+
+  var stream: java.io.OutputStream = {
+    if (useFileOnly)
+      new java.io.FileOutputStream(File.createTempFile("daffodil-", ".tmp", filePath.toFile))
```

Comment:
What happenes if createTempFile fails? Should we throw the new FileIOException, or is whatever exception it throws reasonable?

Reply:
This has been made into a FileIOException